## 0. Load the datasett

In [3]:
import pandas as pd

df_dominant = pd.read_csv('datasett/train.csv')
df_breed = pd.read_csv('datasett/BreedLabels.csv')
df_color = pd.read_csv('datasett/ColorLabels.csv')
df_state = pd.read_csv('datasett/StateLabels.csv')

df_dominant

Type            Name  Age  Breed1  Breed2  Gender  Color1  Color2  \
0         2          Nibble    3     299       0       1       1       7   
1         2     No Name Yet    1     265       0       1       1       2   
2         1          Brisco    1     307       0       1       2       7   
3         1            Miko    4     307       0       2       1       2   
4         1          Hunter    1     307       0       1       1       0   
...     ...             ...  ...     ...     ...     ...     ...     ...   
14988     2             NaN    2     266       0       3       1       0   
14989     2  Serato & Eddie   60     265     264       3       1       4   
14990     2         Monkies    2     265     266       3       5       6   
14991     2         Ms Daym    9     266       0       2       4       7   
14992     1            Fili    1     307     307       1       2       0   

       Color3  MaturitySize  ...  Health  Quantity  Fee  State  \
0           0             1  ...       1         1  100  41326   
1           0             2  ...       1         1    0  41401   
2           0             2  ...       1         1    0  41326   
3           0             2  ...       1         1  150  41401   
4           0             2  ...       1         1    0  41326   
...       ...           ...  ...     ...       ...  ...    ...   
14988       0             2  ...       1         4    0  41326   
14989       7             2  ...       1         2    0  41326   
14990       7             3  ...       1         5   30  41326   
14991       0             1  ...       1         1    0  41336   
14992       0             2  ...       1         1    0  41332   

                              RescuerID  VideoAmt  \
0      8480853f516546f6cf33aa88cd76c379         0   
1      3082c7125d8fb66f7dd4bff4192c8b14         0   
2      fa90fa5b1ee11c86938398b60abc32cb         0   
3      9238e4f44c71a75282e62f7136c6b240         0   
4      95481e953f8aed9ec3d16fc4509537e8         0   
...                                 ...       ...   
14988  61c84bd7bcb6fb31d2d480b1bcf9682e         0   
14989  1d5096c4a5e159a3b750c5cfcf6ceabf         0   
14990  6f40a7acfad5cc0bb3e44591ea446c05         0   
14991  c311c0c569245baa147d91fa4e351ae4         0   
14992  9ed1d5493d223eaa5024c1a031dbc9c2         0   

                                             Description      PetID PhotoAmt  \
0      Nibble is a 3+ month old ball of cuteness. He ...  86e1089a3      1.0   
1      I just found it alone yesterday near my apartm...  6296e909a      2.0   
2      Their pregnant mother was dumped by her irresp...  3422e4906      7.0   
3      Good guard dog, very alert, active, obedience ...  5842f1ff5      8.0   
4      This handsome yet cute boy is up for adoption....  850a43f90      3.0   
...                                                  ...        ...      ...   
14988  I have 4 kittens that need to be adopt urgentl...  dc0935a84      3.0   
14989  Serato(female cat- 3 color) is 4 years old and...  a01ab5b30      3.0   
14990  Mix breed, good temperament kittens. Love huma...  d981b6395      5.0   
14991  she is very shy..adventures and independent..s...  e4da1c9e4      3.0   
14992  Fili just loves laying around and also loves b...  a83d95ead      1.0   

       AdoptionSpeed  
0                  2  
1                  0  
2                  3  
3                  2  
4                  2  
...              ...  
14988              2  
14989              4  
14990              3  
14991              4  
14992              3  

[14993 rows x 24 columns]

In [4]:
df_breed

BreedID  Type         BreedName
0          1     1     Affenpinscher
1          2     1      Afghan Hound
2          3     1  Airedale Terrier
3          4     1            Akbash
4          5     1             Akita
..       ...   ...               ...
302      302     2            Torbie
303      303     2     Tortoiseshell
304      304     2    Turkish Angora
305      305     2       Turkish Van
306      306     2            Tuxedo

[307 rows x 3 columns]

In [5]:
df_color

ColorID ColorName
0        1     Black
1        2     Brown
2        3    Golden
3        4    Yellow
4        5     Cream
5        6      Gray
6        7     White

In [6]:
df_dominant.nunique()

Type                 2
Name              9059
Age                106
Breed1             176
Breed2             135
Gender               3
Color1               7
Color2               7
Color3               6
MaturitySize         4
FurLength            3
Vaccinated           3
Dewormed             3
Sterilized           3
Health               3
Quantity            19
Fee                 74
State               14
RescuerID         5595
VideoAmt             9
Description      14031
PetID            14993
PhotoAmt            31
AdoptionSpeed        5
dtype: int64

## Data preprocessing (state)

- replacing State code to State name

In [7]:
df_state = df_dominant.merge(df_state, left_on='State', right_on='StateID', how='left') # Bring the StateName from df_
df_state

Type            Name  Age  Breed1  Breed2  Gender  Color1  Color2  \
0         2          Nibble    3     299       0       1       1       7   
1         2     No Name Yet    1     265       0       1       1       2   
2         1          Brisco    1     307       0       1       2       7   
3         1            Miko    4     307       0       2       1       2   
4         1          Hunter    1     307       0       1       1       0   
...     ...             ...  ...     ...     ...     ...     ...     ...   
14988     2             NaN    2     266       0       3       1       0   
14989     2  Serato & Eddie   60     265     264       3       1       4   
14990     2         Monkies    2     265     266       3       5       6   
14991     2         Ms Daym    9     266       0       2       4       7   
14992     1            Fili    1     307     307       1       2       0   

       Color3  MaturitySize  ...  Fee  State  \
0           0             1  ...  100  41326   
1           0             2  ...    0  41401   
2           0             2  ...    0  41326   
3           0             2  ...  150  41401   
4           0             2  ...    0  41326   
...       ...           ...  ...  ...    ...   
14988       0             2  ...    0  41326   
14989       7             2  ...    0  41326   
14990       7             3  ...   30  41326   
14991       0             1  ...    0  41336   
14992       0             2  ...    0  41332   

                              RescuerID  VideoAmt  \
0      8480853f516546f6cf33aa88cd76c379         0   
1      3082c7125d8fb66f7dd4bff4192c8b14         0   
2      fa90fa5b1ee11c86938398b60abc32cb         0   
3      9238e4f44c71a75282e62f7136c6b240         0   
4      95481e953f8aed9ec3d16fc4509537e8         0   
...                                 ...       ...   
14988  61c84bd7bcb6fb31d2d480b1bcf9682e         0   
14989  1d5096c4a5e159a3b750c5cfcf6ceabf         0   
14990  6f40a7acfad5cc0bb3e44591ea446c05         0   
14991  c311c0c569245baa147d91fa4e351ae4         0   
14992  9ed1d5493d223eaa5024c1a031dbc9c2         0   

                                             Description      PetID  PhotoAmt  \
0      Nibble is a 3+ month old ball of cuteness. He ...  86e1089a3       1.0   
1      I just found it alone yesterday near my apartm...  6296e909a       2.0   
2      Their pregnant mother was dumped by her irresp...  3422e4906       7.0   
3      Good guard dog, very alert, active, obedience ...  5842f1ff5       8.0   
4      This handsome yet cute boy is up for adoption....  850a43f90       3.0   
...                                                  ...        ...       ...   
14988  I have 4 kittens that need to be adopt urgentl...  dc0935a84       3.0   
14989  Serato(female cat- 3 color) is 4 years old and...  a01ab5b30       3.0   
14990  Mix breed, good temperament kittens. Love huma...  d981b6395       5.0   
14991  she is very shy..adventures and independent..s...  e4da1c9e4       3.0   
14992  Fili just loves laying around and also loves b...  a83d95ead       1.0   

       AdoptionSpeed StateID        StateName  
0                  2   41326         Selangor  
1                  0   41401     Kuala Lumpur  
2                  3   41326         Selangor  
3                  2   41401     Kuala Lumpur  
4                  2   41326         Selangor  
...              ...     ...              ...  
14988              2   41326         Selangor  
14989              4   41326         Selangor  
14990              3   41326         Selangor  
14991              4   41336            Johor  
14992              3   41332  Negeri Sembilan  

[14993 rows x 26 columns]

In [8]:
df_state.nunique()


Type                 2
Name              9059
Age                106
Breed1             176
Breed2             135
Gender               3
Color1               7
Color2               7
Color3               6
MaturitySize         4
FurLength            3
Vaccinated           3
Dewormed             3
Sterilized           3
Health               3
Quantity            19
Fee                 74
State               14
RescuerID         5595
VideoAmt             9
Description      14031
PetID            14993
PhotoAmt            31
AdoptionSpeed        5
StateID             14
StateName           14
dtype: int64

In [9]:
df_state = df_state.drop(columns=['StateID'])

In [10]:
df_state = df_state.rename(columns={'StateName':'State'})


In [11]:
df_state

Type            Name  Age  Breed1  Breed2  Gender  Color1  Color2  \
0         2          Nibble    3     299       0       1       1       7   
1         2     No Name Yet    1     265       0       1       1       2   
2         1          Brisco    1     307       0       1       2       7   
3         1            Miko    4     307       0       2       1       2   
4         1          Hunter    1     307       0       1       1       0   
...     ...             ...  ...     ...     ...     ...     ...     ...   
14988     2             NaN    2     266       0       3       1       0   
14989     2  Serato & Eddie   60     265     264       3       1       4   
14990     2         Monkies    2     265     266       3       5       6   
14991     2         Ms Daym    9     266       0       2       4       7   
14992     1            Fili    1     307     307       1       2       0   

       Color3  MaturitySize  ...  Quantity  Fee  State  \
0           0             1  ...         1  100  41326   
1           0             2  ...         1    0  41401   
2           0             2  ...         1    0  41326   
3           0             2  ...         1  150  41401   
4           0             2  ...         1    0  41326   
...       ...           ...  ...       ...  ...    ...   
14988       0             2  ...         4    0  41326   
14989       7             2  ...         2    0  41326   
14990       7             3  ...         5   30  41326   
14991       0             1  ...         1    0  41336   
14992       0             2  ...         1    0  41332   

                              RescuerID  VideoAmt  \
0      8480853f516546f6cf33aa88cd76c379         0   
1      3082c7125d8fb66f7dd4bff4192c8b14         0   
2      fa90fa5b1ee11c86938398b60abc32cb         0   
3      9238e4f44c71a75282e62f7136c6b240         0   
4      95481e953f8aed9ec3d16fc4509537e8         0   
...                                 ...       ...   
14988  61c84bd7bcb6fb31d2d480b1bcf9682e         0   
14989  1d5096c4a5e159a3b750c5cfcf6ceabf         0   
14990  6f40a7acfad5cc0bb3e44591ea446c05         0   
14991  c311c0c569245baa147d91fa4e351ae4         0   
14992  9ed1d5493d223eaa5024c1a031dbc9c2         0   

                                             Description      PetID  PhotoAmt  \
0      Nibble is a 3+ month old ball of cuteness. He ...  86e1089a3       1.0   
1      I just found it alone yesterday near my apartm...  6296e909a       2.0   
2      Their pregnant mother was dumped by her irresp...  3422e4906       7.0   
3      Good guard dog, very alert, active, obedience ...  5842f1ff5       8.0   
4      This handsome yet cute boy is up for adoption....  850a43f90       3.0   
...                                                  ...        ...       ...   
14988  I have 4 kittens that need to be adopt urgentl...  dc0935a84       3.0   
14989  Serato(female cat- 3 color) is 4 years old and...  a01ab5b30       3.0   
14990  Mix breed, good temperament kittens. Love huma...  d981b6395       5.0   
14991  she is very shy..adventures and independent..s...  e4da1c9e4       3.0   
14992  Fili just loves laying around and also loves b...  a83d95ead       1.0   

      AdoptionSpeed            State  
0                 2         Selangor  
1                 0     Kuala Lumpur  
2                 3         Selangor  
3                 2     Kuala Lumpur  
4                 2         Selangor  
...             ...              ...  
14988             2         Selangor  
14989             4         Selangor  
14990             3         Selangor  
14991             4            Johor  
14992             3  Negeri Sembilan  

[14993 rows x 25 columns]

In [12]:
import os
os.makedirs('output',exist_ok=True)
df_state.to_csv('output/test.csv',index=False)
print("Data Frame has been exported to 'output/test.csv'")

Data Frame has been exported to 'output/test.csv'


## Data Preprocessing (Breed)

In [13]:
df_breed 

BreedID  Type         BreedName
0          1     1     Affenpinscher
1          2     1      Afghan Hound
2          3     1  Airedale Terrier
3          4     1            Akbash
4          5     1             Akita
..       ...   ...               ...
302      302     2            Torbie
303      303     2     Tortoiseshell
304      304     2    Turkish Angora
305      305     2       Turkish Van
306      306     2            Tuxedo

[307 rows x 3 columns]

- to preprocess breed, we need to get breedID and type to match breed name

In [14]:
df_merged_breed = df_dominant.merge(df_breed, left_on=['Type','Breed1'], right_on=['Type','BreedID'], how='left')

In [15]:
df_merged_breed

Type            Name  Age  Breed1  Breed2  Gender  Color1  Color2  \
0         2          Nibble    3     299       0       1       1       7   
1         2     No Name Yet    1     265       0       1       1       2   
2         1          Brisco    1     307       0       1       2       7   
3         1            Miko    4     307       0       2       1       2   
4         1          Hunter    1     307       0       1       1       0   
...     ...             ...  ...     ...     ...     ...     ...     ...   
14988     2             NaN    2     266       0       3       1       0   
14989     2  Serato & Eddie   60     265     264       3       1       4   
14990     2         Monkies    2     265     266       3       5       6   
14991     2         Ms Daym    9     266       0       2       4       7   
14992     1            Fili    1     307     307       1       2       0   

       Color3  MaturitySize  ...  Fee  State  \
0           0             1  ...  100  41326   
1           0             2  ...    0  41401   
2           0             2  ...    0  41326   
3           0             2  ...  150  41401   
4           0             2  ...    0  41326   
...       ...           ...  ...  ...    ...   
14988       0             2  ...    0  41326   
14989       7             2  ...    0  41326   
14990       7             3  ...   30  41326   
14991       0             1  ...    0  41336   
14992       0             2  ...    0  41332   

                              RescuerID  VideoAmt  \
0      8480853f516546f6cf33aa88cd76c379         0   
1      3082c7125d8fb66f7dd4bff4192c8b14         0   
2      fa90fa5b1ee11c86938398b60abc32cb         0   
3      9238e4f44c71a75282e62f7136c6b240         0   
4      95481e953f8aed9ec3d16fc4509537e8         0   
...                                 ...       ...   
14988  61c84bd7bcb6fb31d2d480b1bcf9682e         0   
14989  1d5096c4a5e159a3b750c5cfcf6ceabf         0   
14990  6f40a7acfad5cc0bb3e44591ea446c05         0   
14991  c311c0c569245baa147d91fa4e351ae4         0   
14992  9ed1d5493d223eaa5024c1a031dbc9c2         0   

                                             Description      PetID  PhotoAmt  \
0      Nibble is a 3+ month old ball of cuteness. He ...  86e1089a3       1.0   
1      I just found it alone yesterday near my apartm...  6296e909a       2.0   
2      Their pregnant mother was dumped by her irresp...  3422e4906       7.0   
3      Good guard dog, very alert, active, obedience ...  5842f1ff5       8.0   
4      This handsome yet cute boy is up for adoption....  850a43f90       3.0   
...                                                  ...        ...       ...   
14988  I have 4 kittens that need to be adopt urgentl...  dc0935a84       3.0   
14989  Serato(female cat- 3 color) is 4 years old and...  a01ab5b30       3.0   
14990  Mix breed, good temperament kittens. Love huma...  d981b6395       5.0   
14991  she is very shy..adventures and independent..s...  e4da1c9e4       3.0   
14992  Fili just loves laying around and also loves b...  a83d95ead       1.0   

       AdoptionSpeed BreedID             BreedName  
0                  2   299.0                 Tabby  
1                  0   265.0  Domestic Medium Hair  
2                  3   307.0           Mixed Breed  
3                  2   307.0           Mixed Breed  
4                  2   307.0           Mixed Breed  
...              ...     ...                   ...  
14988              2   266.0   Domestic Short Hair  
14989              4   265.0  Domestic Medium Hair  
14990              3   265.0  Domestic Medium Hair  
14991              4   266.0   Domestic Short Hair  
14992              3   307.0           Mixed Breed  

[14993 rows x 26 columns]

In [16]:
df_merged_breed = df_merged_breed.drop(columns='Type')
df_merged_breed

Name  Age  Breed1  Breed2  Gender  Color1  Color2  Color3  \
0              Nibble    3     299       0       1       1       7       0   
1         No Name Yet    1     265       0       1       1       2       0   
2              Brisco    1     307       0       1       2       7       0   
3                Miko    4     307       0       2       1       2       0   
4              Hunter    1     307       0       1       1       0       0   
...               ...  ...     ...     ...     ...     ...     ...     ...   
14988             NaN    2     266       0       3       1       0       0   
14989  Serato & Eddie   60     265     264       3       1       4       7   
14990         Monkies    2     265     266       3       5       6       7   
14991         Ms Daym    9     266       0       2       4       7       0   
14992            Fili    1     307     307       1       2       0       0   

       MaturitySize  FurLength  ...  Fee  State  \
0                 1          1  ...  100  41326   
1                 2          2  ...    0  41401   
2                 2          2  ...    0  41326   
3                 2          1  ...  150  41401   
4                 2          1  ...    0  41326   
...             ...        ...  ...  ...    ...   
14988             2          2  ...    0  41326   
14989             2          2  ...    0  41326   
14990             3          2  ...   30  41326   
14991             1          1  ...    0  41336   
14992             2          1  ...    0  41332   

                              RescuerID  VideoAmt  \
0      8480853f516546f6cf33aa88cd76c379         0   
1      3082c7125d8fb66f7dd4bff4192c8b14         0   
2      fa90fa5b1ee11c86938398b60abc32cb         0   
3      9238e4f44c71a75282e62f7136c6b240         0   
4      95481e953f8aed9ec3d16fc4509537e8         0   
...                                 ...       ...   
14988  61c84bd7bcb6fb31d2d480b1bcf9682e         0   
14989  1d5096c4a5e159a3b750c5cfcf6ceabf         0   
14990  6f40a7acfad5cc0bb3e44591ea446c05         0   
14991  c311c0c569245baa147d91fa4e351ae4         0   
14992  9ed1d5493d223eaa5024c1a031dbc9c2         0   

                                             Description      PetID  PhotoAmt  \
0      Nibble is a 3+ month old ball of cuteness. He ...  86e1089a3       1.0   
1      I just found it alone yesterday near my apartm...  6296e909a       2.0   
2      Their pregnant mother was dumped by her irresp...  3422e4906       7.0   
3      Good guard dog, very alert, active, obedience ...  5842f1ff5       8.0   
4      This handsome yet cute boy is up for adoption....  850a43f90       3.0   
...                                                  ...        ...       ...   
14988  I have 4 kittens that need to be adopt urgentl...  dc0935a84       3.0   
14989  Serato(female cat- 3 color) is 4 years old and...  a01ab5b30       3.0   
14990  Mix breed, good temperament kittens. Love huma...  d981b6395       5.0   
14991  she is very shy..adventures and independent..s...  e4da1c9e4       3.0   
14992  Fili just loves laying around and also loves b...  a83d95ead       1.0   

      AdoptionSpeed  BreedID             BreedName  
0                 2    299.0                 Tabby  
1                 0    265.0  Domestic Medium Hair  
2                 3    307.0           Mixed Breed  
3                 2    307.0           Mixed Breed  
4                 2    307.0           Mixed Breed  
...             ...      ...                   ...  
14988             2    266.0   Domestic Short Hair  
14989             4    265.0  Domestic Medium Hair  
14990             3    265.0  Domestic Medium Hair  
14991             4    266.0   Domestic Short Hair  
14992             3    307.0           Mixed Breed  

[14993 rows x 25 columns]

In [17]:
null_count = df_merged_breed['BreedName'].isnull().sum()
print(null_count)

17


since only 17 out of 14992 (0.11%) is null, i will just drop it

In [18]:
df_merged_breed = df_merged_breed.dropna(subset='BreedName')
null_count = df_merged_breed['BreedName'].isnull().sum()
print(null_count)

0


In [19]:
import os
os.makedirs('output', exist_ok=True)
df_merged_breed.to_csv('output/breed.csv', index=False)

# changing gender. 1 male, 2 female, 3 mixed 


In [20]:
df_gender = df_merged_breed
df_gender['Gender'] = df_gender['Gender'].replace({1:'Male',2:'Female',3:'Mixed'})
df_gender

C:\Users\User\AppData\Local\Temp\ipykernel_11744\3239459164.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_gender['Gender'] = df_gender['Gender'].replace({1:'Male',2:'Female',3:'Mixed'})


Name  Age  Breed1  Breed2  Gender  Color1  Color2  Color3  \
0              Nibble    3     299       0    Male       1       7       0   
1         No Name Yet    1     265       0    Male       1       2       0   
2              Brisco    1     307       0    Male       2       7       0   
3                Miko    4     307       0  Female       1       2       0   
4              Hunter    1     307       0    Male       1       0       0   
...               ...  ...     ...     ...     ...     ...     ...     ...   
14988             NaN    2     266       0   Mixed       1       0       0   
14989  Serato & Eddie   60     265     264   Mixed       1       4       7   
14990         Monkies    2     265     266   Mixed       5       6       7   
14991         Ms Daym    9     266       0  Female       4       7       0   
14992            Fili    1     307     307    Male       2       0       0   

       MaturitySize  FurLength  ...  Fee  State  \
0                 1          1  ...  100  41326   
1                 2          2  ...    0  41401   
2                 2          2  ...    0  41326   
3                 2          1  ...  150  41401   
4                 2          1  ...    0  41326   
...             ...        ...  ...  ...    ...   
14988             2          2  ...    0  41326   
14989             2          2  ...    0  41326   
14990             3          2  ...   30  41326   
14991             1          1  ...    0  41336   
14992             2          1  ...    0  41332   

                              RescuerID  VideoAmt  \
0      8480853f516546f6cf33aa88cd76c379         0   
1      3082c7125d8fb66f7dd4bff4192c8b14         0   
2      fa90fa5b1ee11c86938398b60abc32cb         0   
3      9238e4f44c71a75282e62f7136c6b240         0   
4      95481e953f8aed9ec3d16fc4509537e8         0   
...                                 ...       ...   
14988  61c84bd7bcb6fb31d2d480b1bcf9682e         0   
14989  1d5096c4a5e159a3b750c5cfcf6ceabf         0   
14990  6f40a7acfad5cc0bb3e44591ea446c05         0   
14991  c311c0c569245baa147d91fa4e351ae4         0   
14992  9ed1d5493d223eaa5024c1a031dbc9c2         0   

                                             Description      PetID  PhotoAmt  \
0      Nibble is a 3+ month old ball of cuteness. He ...  86e1089a3       1.0   
1      I just found it alone yesterday near my apartm...  6296e909a       2.0   
2      Their pregnant mother was dumped by her irresp...  3422e4906       7.0   
3      Good guard dog, very alert, active, obedience ...  5842f1ff5       8.0   
4      This handsome yet cute boy is up for adoption....  850a43f90       3.0   
...                                                  ...        ...       ...   
14988  I have 4 kittens that need to be adopt urgentl...  dc0935a84       3.0   
14989  Serato(female cat- 3 color) is 4 years old and...  a01ab5b30       3.0   
14990  Mix breed, good temperament kittens. Love huma...  d981b6395       5.0   
14991  she is very shy..adventures and independent..s...  e4da1c9e4       3.0   
14992  Fili just loves laying around and also loves b...  a83d95ead       1.0   

      AdoptionSpeed  BreedID             BreedName  
0                 2    299.0                 Tabby  
1                 0    265.0  Domestic Medium Hair  
2                 3    307.0           Mixed Breed  
3                 2    307.0           Mixed Breed  
4                 2    307.0           Mixed Breed  
...             ...      ...                   ...  
14988             2    266.0   Domestic Short Hair  
14989             4    265.0  Domestic Medium Hair  
14990             3    265.0  Domestic Medium Hair  
14991             4    266.0   Domestic Short Hair  
14992             3    307.0           Mixed Breed  

[14976 rows x 25 columns]

In [21]:
# checking for issues
df_gender.nunique()

Name              9046
Age                106
Breed1             174
Breed2             135
Gender               3
Color1               7
Color2               7
Color3               6
MaturitySize         4
FurLength            3
Vaccinated           3
Dewormed             3
Sterilized           3
Health               3
Quantity            19
Fee                 74
State               14
RescuerID         5590
VideoAmt             9
Description      14017
PetID            14976
PhotoAmt            31
AdoptionSpeed        5
BreedID            174
BreedName          174
dtype: int64

gender settled, no worries

In [22]:
import os
os.makedirs('output',exist_ok=True)
df_gender.to_csv('output/gender.csv', index=True)

PermissionError: [Errno 13] Permission denied: 'output/gender.csv'

accidentally dropped the Type, merging back from dominant(master df)

In [25]:
df_color = df_gender
color_mapping = {1:'Black',2:'Brown',3:'Golden',4:'Yellow',5:'Cream',6:'Gray',7:'White'}

df_color[['Color1','Color2','Color3']] = df_color[['Color1','Color2','Color3']].replace(color_mapping)
df_color

C:\Users\User\AppData\Local\Temp\ipykernel_11744\581588323.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_color[['Color1','Color2','Color3']] = df_color[['Color1','Color2','Color3']].replace(color_mapping)


Name  Age  Breed1  Breed2  Gender  Color1  Color2 Color3  \
0              Nibble    3     299       0    Male   Black   White      0   
1         No Name Yet    1     265       0    Male   Black   Brown      0   
2              Brisco    1     307       0    Male   Brown   White      0   
3                Miko    4     307       0  Female   Black   Brown      0   
4              Hunter    1     307       0    Male   Black       0      0   
...               ...  ...     ...     ...     ...     ...     ...    ...   
14988             NaN    2     266       0   Mixed   Black       0      0   
14989  Serato & Eddie   60     265     264   Mixed   Black  Yellow  White   
14990         Monkies    2     265     266   Mixed   Cream    Gray  White   
14991         Ms Daym    9     266       0  Female  Yellow   White      0   
14992            Fili    1     307     307    Male   Brown       0      0   

       MaturitySize  FurLength  ...  Fee  State  \
0                 1          1  ...  100  41326   
1                 2          2  ...    0  41401   
2                 2          2  ...    0  41326   
3                 2          1  ...  150  41401   
4                 2          1  ...    0  41326   
...             ...        ...  ...  ...    ...   
14988             2          2  ...    0  41326   
14989             2          2  ...    0  41326   
14990             3          2  ...   30  41326   
14991             1          1  ...    0  41336   
14992             2          1  ...    0  41332   

                              RescuerID  VideoAmt  \
0      8480853f516546f6cf33aa88cd76c379         0   
1      3082c7125d8fb66f7dd4bff4192c8b14         0   
2      fa90fa5b1ee11c86938398b60abc32cb         0   
3      9238e4f44c71a75282e62f7136c6b240         0   
4      95481e953f8aed9ec3d16fc4509537e8         0   
...                                 ...       ...   
14988  61c84bd7bcb6fb31d2d480b1bcf9682e         0   
14989  1d5096c4a5e159a3b750c5cfcf6ceabf         0   
14990  6f40a7acfad5cc0bb3e44591ea446c05         0   
14991  c311c0c569245baa147d91fa4e351ae4         0   
14992  9ed1d5493d223eaa5024c1a031dbc9c2         0   

                                             Description      PetID  PhotoAmt  \
0      Nibble is a 3+ month old ball of cuteness. He ...  86e1089a3       1.0   
1      I just found it alone yesterday near my apartm...  6296e909a       2.0   
2      Their pregnant mother was dumped by her irresp...  3422e4906       7.0   
3      Good guard dog, very alert, active, obedience ...  5842f1ff5       8.0   
4      This handsome yet cute boy is up for adoption....  850a43f90       3.0   
...                                                  ...        ...       ...   
14988  I have 4 kittens that need to be adopt urgentl...  dc0935a84       3.0   
14989  Serato(female cat- 3 color) is 4 years old and...  a01ab5b30       3.0   
14990  Mix breed, good temperament kittens. Love huma...  d981b6395       5.0   
14991  she is very shy..adventures and independent..s...  e4da1c9e4       3.0   
14992  Fili just loves laying around and also loves b...  a83d95ead       1.0   

      AdoptionSpeed  BreedID             BreedName  
0                 2    299.0                 Tabby  
1                 0    265.0  Domestic Medium Hair  
2                 3    307.0           Mixed Breed  
3                 2    307.0           Mixed Breed  
4                 2    307.0           Mixed Breed  
...             ...      ...                   ...  
14988             2    266.0   Domestic Short Hair  
14989             4    265.0  Domestic Medium Hair  
14990             3    265.0  Domestic Medium Hair  
14991             4    266.0   Domestic Short Hair  
14992             3    307.0           Mixed Breed  

[14976 rows x 25 columns]

In [26]:
# check for issue
zero_count = (df_color['Color1'] == 0).sum()
print(zero_count)
zero_count = (df_color['Color2'] == 0).sum()
print(zero_count)
zero_count = (df_color['Color3'] == 0).sum()
print(zero_count)

0
4468
10594


In [28]:
# apply color1 to color2
df_color['Color2'] = df_color.apply(lambda row: row['Color1'] if row['Color2'] == 0 else row['Color2'], axis=1)
df_color['Color3'] = df_color.apply(lambda row: row['Color2'] if row['Color3'] == 0 else row['Color3'], axis=1)
df_color

C:\Users\User\AppData\Local\Temp\ipykernel_11744\3417598940.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_color['Color2'] = df_color.apply(lambda row: row['Color1'] if row['Color2'] == 0 else row['Color2'], axis=1)
C:\Users\User\AppData\Local\Temp\ipykernel_11744\3417598940.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_color['Color3'] = df_color.apply(lambda row: row['Color2'] if row['Color3'] == 0 else row['Color3'], axis=1)


Name  Age  Breed1  Breed2  Gender  Color1  Color2 Color3  \
0              Nibble    3     299       0    Male   Black   White  White   
1         No Name Yet    1     265       0    Male   Black   Brown  Brown   
2              Brisco    1     307       0    Male   Brown   White  White   
3                Miko    4     307       0  Female   Black   Brown  Brown   
4              Hunter    1     307       0    Male   Black   Black  Black   
...               ...  ...     ...     ...     ...     ...     ...    ...   
14988             NaN    2     266       0   Mixed   Black   Black  Black   
14989  Serato & Eddie   60     265     264   Mixed   Black  Yellow  White   
14990         Monkies    2     265     266   Mixed   Cream    Gray  White   
14991         Ms Daym    9     266       0  Female  Yellow   White  White   
14992            Fili    1     307     307    Male   Brown   Brown  Brown   

       MaturitySize  FurLength  ...  Fee  State  \
0                 1          1  ...  100  41326   
1                 2          2  ...    0  41401   
2                 2          2  ...    0  41326   
3                 2          1  ...  150  41401   
4                 2          1  ...    0  41326   
...             ...        ...  ...  ...    ...   
14988             2          2  ...    0  41326   
14989             2          2  ...    0  41326   
14990             3          2  ...   30  41326   
14991             1          1  ...    0  41336   
14992             2          1  ...    0  41332   

                              RescuerID  VideoAmt  \
0      8480853f516546f6cf33aa88cd76c379         0   
1      3082c7125d8fb66f7dd4bff4192c8b14         0   
2      fa90fa5b1ee11c86938398b60abc32cb         0   
3      9238e4f44c71a75282e62f7136c6b240         0   
4      95481e953f8aed9ec3d16fc4509537e8         0   
...                                 ...       ...   
14988  61c84bd7bcb6fb31d2d480b1bcf9682e         0   
14989  1d5096c4a5e159a3b750c5cfcf6ceabf         0   
14990  6f40a7acfad5cc0bb3e44591ea446c05         0   
14991  c311c0c569245baa147d91fa4e351ae4         0   
14992  9ed1d5493d223eaa5024c1a031dbc9c2         0   

                                             Description      PetID  PhotoAmt  \
0      Nibble is a 3+ month old ball of cuteness. He ...  86e1089a3       1.0   
1      I just found it alone yesterday near my apartm...  6296e909a       2.0   
2      Their pregnant mother was dumped by her irresp...  3422e4906       7.0   
3      Good guard dog, very alert, active, obedience ...  5842f1ff5       8.0   
4      This handsome yet cute boy is up for adoption....  850a43f90       3.0   
...                                                  ...        ...       ...   
14988  I have 4 kittens that need to be adopt urgentl...  dc0935a84       3.0   
14989  Serato(female cat- 3 color) is 4 years old and...  a01ab5b30       3.0   
14990  Mix breed, good temperament kittens. Love huma...  d981b6395       5.0   
14991  she is very shy..adventures and independent..s...  e4da1c9e4       3.0   
14992  Fili just loves laying around and also loves b...  a83d95ead       1.0   

      AdoptionSpeed  BreedID             BreedName  
0                 2    299.0                 Tabby  
1                 0    265.0  Domestic Medium Hair  
2                 3    307.0           Mixed Breed  
3                 2    307.0           Mixed Breed  
4                 2    307.0           Mixed Breed  
...             ...      ...                   ...  
14988             2    266.0   Domestic Short Hair  
14989             4    265.0  Domestic Medium Hair  
14990             3    265.0  Domestic Medium Hair  
14991             4    266.0   Domestic Short Hair  
14992             3    307.0           Mixed Breed  

[14976 rows x 25 columns]

In [29]:
import os
os.makedirs('output',exist_ok=True)
df_color.to_csv('output/color.csv',index=True)

MaturitySize - Size at maturity (1 = Small, 2 = Medium, 3 = Large, 4 = Extra Large, 0 = Not Specified)


In [30]:
df_maturity = df_color


In [31]:
df_maturity['MaturitySize'] = df_maturity['MaturitySize'].replace({0:'Not Specified',1:'Small',2:'Medium',3:'Large',4:'Extra Large'})
df_maturity

C:\Users\User\AppData\Local\Temp\ipykernel_11744\2119530153.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_maturity['MaturitySize'] = df_maturity['MaturitySize'].replace({0:'Not Specified',1:'Small',2:'Medium',3:'Large',4:'Extra Large'})


Name  Age  Breed1  Breed2  Gender  Color1  Color2 Color3  \
0              Nibble    3     299       0    Male   Black   White  White   
1         No Name Yet    1     265       0    Male   Black   Brown  Brown   
2              Brisco    1     307       0    Male   Brown   White  White   
3                Miko    4     307       0  Female   Black   Brown  Brown   
4              Hunter    1     307       0    Male   Black   Black  Black   
...               ...  ...     ...     ...     ...     ...     ...    ...   
14988             NaN    2     266       0   Mixed   Black   Black  Black   
14989  Serato & Eddie   60     265     264   Mixed   Black  Yellow  White   
14990         Monkies    2     265     266   Mixed   Cream    Gray  White   
14991         Ms Daym    9     266       0  Female  Yellow   White  White   
14992            Fili    1     307     307    Male   Brown   Brown  Brown   

      MaturitySize  FurLength  ...  Fee  State  \
0            Small          1  ...  100  41326   
1           Medium          2  ...    0  41401   
2           Medium          2  ...    0  41326   
3           Medium          1  ...  150  41401   
4           Medium          1  ...    0  41326   
...            ...        ...  ...  ...    ...   
14988       Medium          2  ...    0  41326   
14989       Medium          2  ...    0  41326   
14990        Large          2  ...   30  41326   
14991        Small          1  ...    0  41336   
14992       Medium          1  ...    0  41332   

                              RescuerID  VideoAmt  \
0      8480853f516546f6cf33aa88cd76c379         0   
1      3082c7125d8fb66f7dd4bff4192c8b14         0   
2      fa90fa5b1ee11c86938398b60abc32cb         0   
3      9238e4f44c71a75282e62f7136c6b240         0   
4      95481e953f8aed9ec3d16fc4509537e8         0   
...                                 ...       ...   
14988  61c84bd7bcb6fb31d2d480b1bcf9682e         0   
14989  1d5096c4a5e159a3b750c5cfcf6ceabf         0   
14990  6f40a7acfad5cc0bb3e44591ea446c05         0   
14991  c311c0c569245baa147d91fa4e351ae4         0   
14992  9ed1d5493d223eaa5024c1a031dbc9c2         0   

                                             Description      PetID  PhotoAmt  \
0      Nibble is a 3+ month old ball of cuteness. He ...  86e1089a3       1.0   
1      I just found it alone yesterday near my apartm...  6296e909a       2.0   
2      Their pregnant mother was dumped by her irresp...  3422e4906       7.0   
3      Good guard dog, very alert, active, obedience ...  5842f1ff5       8.0   
4      This handsome yet cute boy is up for adoption....  850a43f90       3.0   
...                                                  ...        ...       ...   
14988  I have 4 kittens that need to be adopt urgentl...  dc0935a84       3.0   
14989  Serato(female cat- 3 color) is 4 years old and...  a01ab5b30       3.0   
14990  Mix breed, good temperament kittens. Love huma...  d981b6395       5.0   
14991  she is very shy..adventures and independent..s...  e4da1c9e4       3.0   
14992  Fili just loves laying around and also loves b...  a83d95ead       1.0   

      AdoptionSpeed  BreedID             BreedName  
0                 2    299.0                 Tabby  
1                 0    265.0  Domestic Medium Hair  
2                 3    307.0           Mixed Breed  
3                 2    307.0           Mixed Breed  
4                 2    307.0           Mixed Breed  
...             ...      ...                   ...  
14988             2    266.0   Domestic Short Hair  
14989             4    265.0  Domestic Medium Hair  
14990             3    265.0  Domestic Medium Hair  
14991             4    266.0   Domestic Short Hair  
14992             3    307.0           Mixed Breed  

[14976 rows x 25 columns]

In [32]:
import os
os.makedirs('output',exist_ok=True)
df_maturity.to_csv('output/maturity.csv',index=True)

In [34]:
print(df_maturity.nunique())


Name              9046
Age                106
Breed1             174
Breed2             135
Gender               3
Color1               7
Color2               7
Color3               7
MaturitySize         4
FurLength            3
Vaccinated           3
Dewormed             3
Sterilized           3
Health               3
Quantity            19
Fee                 74
State               14
RescuerID         5590
VideoAmt             9
Description      14017
PetID            14976
PhotoAmt            31
AdoptionSpeed        5
BreedID            174
BreedName          174
dtype: int64
